homework_2:

Arthasmethil opened the pull request:
pom was changed,
xml file for hw 2 was created
exercise 1: ExerciseSoftAssertsTest.java was created
exercise 2: ExerciseAssertJTest.java was created
utils: LogIn.java, RegexChecker.java and SeleniumAbstractCore.java were created
constants: ConstantsForUserAndTests.java was created

Arthasmethil committed:
homework_2:
pom was changed,
xml file for hw 2 was created
exercise 1: ExerciseSoftAssertsTest.java was created
exercise 2: ExerciseAssertJTest.java was created
utils: LogIn.java, RegexChecker.java and SeleniumAbstractCore.java were created
constants: ConstantsForUserAndTests.java was created

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/constants/ConstantsForUserAndTests.java:
```diff
@@ -0,0 +1,15 @@
+    public static final String EXPECTED_ICON_TEXT_TWO = "To be flexible and\ncustomizable";
+    public static final String EXPECTED_ICON_TEXT_THREE = "To be multiplatform";
+    public static final String EXPECTED_ICON_TEXT_FOUR = "Already have good base"
+        + "\n(about 20 internal and\nsome external projects),\nwish to get more…";
```
Lets replace these by list of expected texts, something like `psf List<String> EXPECTED_ICON_TEXTS = List.of("veni", "vidi", "vici");`

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.testng.annotations.Test;
+import org.testng.asserts.SoftAssert;
```
Please use AssertJ analogue
```suggestion
import org.assertj.core.api.SoftAssertions;
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+
+    @Test
+    public void checkAndSearchElementsOnPageWithSoftAsserts() {
+        SoftAssert softAssert = new SoftAssert();
```
To fluent read variable name `softAssert` can be replaced by `softly`, to have something like `softly.assertThat(...)`
```suggestion
        SoftAssertions softly = new SoftAssertions();
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+        // 1. Open test site by URL
+        driver.get(ConstantsForUserAndTests.SITE_ADDRESS);
+        // 2. Assert Browser title
+        softAssert.assertEquals(driver.getTitle(), "Home Page");
```
Expected icon texts are stored in separate place, lets move other expected values to the same place too

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+            .visibilityOfElementLocated(
+                By.cssSelector(".m-l8 > li:nth-child(4) > a")));
+        softAssert.assertTrue(metalsAndColorsHeader.isDisplayed());
+        softAssert.assertEquals(metalsAndColorsHeader.getText(), "METALS & COLORS");
```
1. `wait` here looks unnecessary because you check static element, and it's content does not depend on previous actions
1. use one selector that extracts list of top level items of header menu 
1. put expected texts in a list (see my comment for class with constants
1. AssertJ is good for fluent asserts, so all required checks can be performed in one `assertThat` statement (ping me in teams, I'll explain with examples)

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+        // 4. Name is displayed and equals to expected result
+        softAssert.assertTrue(wait.until(ExpectedConditions
+                                      .visibilityOfElementLocated(By.id("user-name")))
+                                  .isDisplayed());
```
`.visibilityOf...` already checks `.isDisplayed`. In other words you say "find me BUTTER" and after that check "BUTTER == BUTTER", which looks a bit weird

Log in step is crucial here, so if this step fails (which will happen if NoSuchElementException is thrown) it is good behavior, as other steps are not informative

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+        softAssert.assertEquals(metalsAndColorsHeader.getText(), "METALS & COLORS");
+        // 6. Assert that there are 4 images on the Index Page and they are displayed
+        List<WebElement> benefitIcons = wait.until(ExpectedConditions
+            .visibilityOfAllElementsLocatedBy(By.className("benefit-icon")));
```
please replace
```suggestion
            .visibilityOfAllElementsLocatedBy(By.cssSelector(".benefit-icon")));
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+            .visibilityOfAllElementsLocatedBy(By.className("benefit-icon")));
+        for (WebElement icon : benefitIcons) {
+            softAssert.assertTrue(icon.isDisplayed());
+        }
```
1. same as in one of previous comment BUTTER == BUTTER
1. such loops in test are not a good practice, use dedicated collections assert (e.g. `.allMatch`, `.allSatisfy`)
1. number of elements is not checked here

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+        softAssert.assertEquals(benefitText.get(0).getText(), ConstantsForUserAndTests.EXPECTED_ICON_TEXT_ONE);
+        softAssert.assertEquals(benefitText.get(1).getText(), ConstantsForUserAndTests.EXPECTED_ICON_TEXT_TWO);
+        softAssert.assertEquals(benefitText.get(2).getText(), ConstantsForUserAndTests.EXPECTED_ICON_TEXT_THREE);
+        softAssert.assertEquals(benefitText.get(3).getText(), ConstantsForUserAndTests.EXPECTED_ICON_TEXT_FOUR);
```
replace this by one assert (ping me in teams, I'll give example)

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+        // 8. Assert that there is the iframe with “Frame Button” exist
+        softAssert.assertTrue(wait.until(ExpectedConditions
+            .visibilityOfElementLocated(By.id("frame")))
+              .isDisplayed());
```
If this element is not present test fails (exception is raised). Please replace it by `.findElements(...)` and check size+isDisplayed of resulting list

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+        driver.switchTo().frame("frame");
+        softAssert.assertTrue(wait.until(ExpectedConditions
+                                      .visibilityOfElementLocated(By.id("frame-button")))
+                                  .isDisplayed());
```
same as previous comment

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,107 @@
+        WebElement elementsPacksLeftBar = wait.until(ExpectedConditions
+            .visibilityOfElementLocated(By.xpath("//span[text()='Elements packs']")));
+        softAssert.assertTrue(elementsPacksLeftBar.isDisplayed());
+        softAssert.assertEquals(elementsPacksLeftBar.getText(), "Elements packs");
```
same as one of previous comment
1. use one selector returning collection
1. check using one assert statement (collections of texts equality and all isDisplayed)

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex2/ExerciseAssertJTest.java:
```diff
@@ -0,0 +1,71 @@
+        // 9.1 for each checkbox there is an individual log row and value is corresponded to the status of checkbox
+        //
+        WebElement waterLog =
+            wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//li[contains(text(), 'Water')]")));
+        WebElement windLog =
+            wait.until(ExpectedConditions.visibilityOfElementLocated(By.xpath("//li[contains(text(), 'Wind')]")));
+        Assertions.assertThat(RegexChecker.cutTimeOfLogString(waterLog.getText()))
+                  .isEqualTo("Water: condition changed to true");
+        Assertions.assertThat(RegexChecker.cutTimeOfLogString(windLog.getText()))
+                  .isEqualTo("Wind: condition changed to true");
```
To avoid Butter=Butter situation I suggest to get all lines from log-element, trim timestamp and check that collection contains some strings

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/utils/LogIn.java:
```diff
@@ -0,0 +1,39 @@
+        sendLoginName(wait, ConstantsForUserAndTests.USERNAME);
+        sendLoginPass(wait, ConstantsForUserAndTests.PASS);
+        logInSubmit(wait);
+    }
```
Would be better to pass username/password as arguments (in specific class, e.g. User(login, password, displayedName))

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/utils/RegexChecker.java:
```diff
@@ -0,0 +1,16 @@
+public class RegexChecker {
+
+    public static String cutTimeOfLogString(String logStringWithTime) {
+        Matcher matcher = Pattern.compile("[a-zA-Z]+:.*").matcher(logStringWithTime);
+        StringBuilder stringBuilder = new StringBuilder();
+        while (matcher.find()) {
+            stringBuilder.append(matcher.group());
+        }
+        return stringBuilder.toString();
```
Lets simplify this code
```suggestion
        return logStringWithTime.replaceFirst("^[\\d]{2}:[\\d]{2}:[\\d]{2} ", "");
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/utils/SeleniumAbstractCore.java:
```diff
@@ -0,0 +1,33 @@
+
+    @BeforeClass
+    public void setUp() {
+        driver = new ChromeDriver();
+        wait = new WebDriverWait(driver, Duration.ofSeconds(10));
+        driver.manage().window().maximize();
```
I suggest to start browser already maximized. Please see second option in article https://unadkatjash.medium.com/maximizing-chrome-window-in-selenium-webdriver-using-java-73e20360f2d3

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/utils/RegexChecker.java:
```diff
@@ -0,0 +1,16 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class RegexChecker {
```
Class name does not reflect its purpose

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/utils/LogIn.java:
```diff
@@ -0,0 +1,39 @@
+package com.epam.tc.hw2.utils;
```
Please move to steps package
```suggestion
package com.epam.tc.hw2.steps;
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/utils/LogIn.java:
```diff
@@ -0,0 +1,39 @@
+import org.openqa.selenium.support.ui.ExpectedConditions;
+import org.openqa.selenium.support.ui.WebDriverWait;
+
+public class LogIn {
```
Please rename 
```suggestion
public class LogInSteps {
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex2/ExerciseAssertJTest.java:
```diff
@@ -0,0 +1,71 @@
+public class ExerciseAssertJTest extends SeleniumAbstractCore {
+
+    @Test
+    public void checkAndSearchElementsOnIndexPage() {
```
Test checks elements not on index page, please rename method

shoggy reviewed (changes requested):
Do not hesitate to contact me in teams

## Test structure requirements:
- [x] Browser: chrome
- [x] Window: maximized
- [x] WebElement locators
  - [x] should not be autogenerated
  - [x] follow priority
    - DOM locator
    - CSS selector
    - XPath selector
- [ ] Proper waits
- [x] Block starts by comment line (e.g. `// 2. Assert Browser title`)
- [x] Code conforms to the Java Code Convention
- [x] Checkstyle plugin does not display errors
- [x] TestNG is used as unit test framework
- [x] TestNG is configured using xml file
- [x] Dedicated Maven profile is created
- [x] AssertJ is used instead of basic TestNG asserts
- [x] Tests are ran using maven commands

## File placement requirements:
- [x] Classes should be placed into `src/test/java/com/epam/tc/hw2` package and sub-packages
- [x] Homework should be developed in new Java classes
- [x] Each exercise is placed in dedicated package
- [x] Test methods' names should reflect checking functionality

## Exercises
- [x] Exercise 1
- [x] Exercise 2

Arthasmethil committed:
Scope of work:
-Rename LogInSteps class,
-Rename checkAndSearchElementsOnDifferentElementsPage method
-Update ConstantsForUserAndTests
-Update SeleniumAbstractCore
-Refactor checkAndSearchElementsOnPageWithSoftAsserts

Arthasmethil committed:
Scope of work:
-Change package for StringEditorForLogs,
-Refactor assertions with collections
-Refactor LogInSteps
-Remove unnecessary waits

Arthasmethil committed:
Scope of work:
-Separate User data from constants,
-Refactor step with iframe element

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,72 @@
+              .isEqualTo(User.EXPECTED_DISPLAYED_NAME);
+        // 5. Assert that there are 4 items on the header section are displayed and they have proper texts
+        List<WebElement> headerMenu = driver.findElements(
+            By.xpath("//ul[@class='uui-navigation nav navbar-nav m-l8']/li"));
```
```suggestion
            By.css("ul.uui-navigation.nav.navbar-nav.m-l8 > li"));
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex1/ExerciseSoftAssertsTest.java:
```diff
@@ -0,0 +1,72 @@
+        driver.switchTo().defaultContent();
+        // 11. Assert that there are 5 items in the Left Section are displayed and they have proper text
+        List<WebElement> homeLeftBar = driver.findElements(
+            By.xpath("//*[@id='mCSB_1_container']/ul/*"));
```
something like
```suggestion
            By.css("#mCSB_1_container > ul > *"));
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex2/ExerciseAssertJTest.java:
```diff
@@ -0,0 +1,60 @@
+                  .isEqualTo(User.EXPECTED_DISPLAYED_NAME);
+        // 5. Open through the header menu Service -> Different Elements Page
+        // 5.1 Dropdown caret is opened
+        driver.findElement(By.xpath("//li[@class='dropdown']")).click();
```
```suggestion
        driver.findElement(By.css("li.dropdown")).click();
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/ex2/ExerciseAssertJTest.java:
```diff
@@ -0,0 +1,60 @@
+        // 9. Assert that for each checkbox there is an individual log row and value
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.className("info-panel-section")));
+        List<String> logStringWithoutTime = driver
+            .findElements(By.xpath("//ul[@class='panel-body-list logs']/li"))
```
```suggestion
            .findElements(By.css("ul.panel-body-list.logs > li"))
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/steps/LogInSteps.java:
```diff
@@ -0,0 +1,38 @@
+
+
+    public static void openLoginCaret(WebDriver driver) {
+        driver.findElement(By.cssSelector("li[class*='uui-profile-menu']")).click();
```
```suggestion
        driver.findElement(By.cssSelector("li.uui-profile-menu")).click();
```

shoggy commented on home-tasks/src/test/java/com/epam/tc/hw2/steps/LogInSteps.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    public static void visibilityOfLoginDropdownMenu(WebDriverWait wait) {
+        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("div[class*='dropdown-menu-login']")));
```
```suggestion
        wait.until(ExpectedConditions.visibilityOfElementLocated(By.cssSelector("div.dropdown-menu-login")));
```

shoggy reviewed (changes requested):
Looks good
Update locator according to type priority: DOM, CSS, XPath

Arthasmethil committed:
Scope of work:
-Refactor selectors
